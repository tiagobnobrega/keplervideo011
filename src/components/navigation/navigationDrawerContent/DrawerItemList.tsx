import {
  HWEvent,
  TVFocusGuideView,
  useTVEventHandler,
} from '@amzn/react-native-kepler';
import {
  DrawerDescriptorMap,
  DrawerNavigationHelpers,
} from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import {
  DrawerNavigationState,
  ParamListBase,
} from '@amzn/react-navigation__native';
import {
  NavigationState,
  PartialState,
  Route,
} from '@amzn/react-navigation__routers/lib/typescript/src/types';
import { isEqual } from 'lodash';
import React, { useEffect, useState } from 'react';
import { findNodeHandle, StyleSheet, TouchableOpacity } from 'react-native';
import { useSelector } from 'react-redux';
import { DPADEventType, EVENT_KEY_DOWN } from '../../../constants';
import { focusSelectors } from '../../../store/focus/focusSlice';
import { recordCustomEvent } from '../../../utils/helperNewRelic';
import { Screens } from '../types';
import DrawerItem from './DrawerItem';

type NavigationRoute<
  ParamList extends ParamListBase,
  RouteName extends keyof ParamList,
> = Route<Extract<RouteName, string>, ParamList[RouteName]> & {
  state?: NavigationState | PartialState<NavigationState>;
};

export interface DrawerItemListProps {
  state: DrawerNavigationState<ParamListBase>;
  navigation: DrawerNavigationHelpers;
  descriptors: DrawerDescriptorMap;
  isDrawerInFocus: boolean;
  onDrawerListFocus: () => void;
  onDrawerListBlur: () => void;
  drawerWidth: number;
  testID?: string;
}

const NOT_FOCUSED_ITEM = -1;

const DrawerItemList = ({
  state,
  navigation,
  descriptors,
  onDrawerListFocus,
  onDrawerListBlur,
  isDrawerInFocus,
  drawerWidth,
  testID,
}: DrawerItemListProps) => {
  const [rootNodeHandle, setRootNodeHandle] = useState<number | undefined>(
    undefined,
  );

  const [selectedItem, setSelectedItem] = useState(0);
  const [focusedItem, setFocusedItem] = useState(NOT_FOCUSED_ITEM);
  const currentFocusScreen = useSelector(focusSelectors.getCurrentFocus);

  useTVEventHandler((evt: HWEvent) => {
    if (!isDrawerInFocus) {
      return;
    }

    if (evt?.eventKeyAction !== EVENT_KEY_DOWN) {
      return;
    }

    if (evt?.eventType === DPADEventType.SELECT) {
      onDrawerListBlur();
    }
    if (evt.eventType === DPADEventType.UP && focusedItem > 0) {
      setFocusedItem(focusedItem - 1);
    } else if (
      evt.eventType === DPADEventType.DOWN &&
      focusedItem < state.routes.length - 1
    ) {
      setFocusedItem(focusedItem + 1);
    } else if (evt.eventType === DPADEventType.SELECT) {
      const { name } = state.routes[focusedItem];
      onItemPress(name);
    }
  });

  useEffect(() => {
    if (currentFocusScreen === Screens.DEFAULT_SCREEN) {
      setFocusedItem(0);
      setSelectedItem(0);
    }
  }, [currentFocusScreen]);

  useEffect(() => {
    if (!isDrawerInFocus) {
      setFocusedItem(NOT_FOCUSED_ITEM);
    }
  }, [isDrawerInFocus]);

  const onItemPress = (routeName: string) => {
    const index: number = state.routes.findIndex(r => r.name === routeName);
    setFocusedItem(index);
    setSelectedItem(index);
    navigation.navigate(routeName);
    recordCustomEvent('Navigated: ', {
      name: routeName,
    });
  };

  const onFocus = () => {
    setFocusedItem(selectedItem);
    onDrawerListFocus();
  };

  const onBlur = () => {
    setFocusedItem(NOT_FOCUSED_ITEM);
    onDrawerListBlur();
  };

  return (
    <TVFocusGuideView style={styles.listContainer} trapFocusDown>
      <TouchableOpacity
        activeOpacity={1}
        ref={component => {
          const nodeHandle = findNodeHandle(component);
          setRootNodeHandle(nodeHandle ? nodeHandle : undefined);
        }}
        style={styles.listContainer}
        nextFocusDown={rootNodeHandle}
        nextFocusUp={rootNodeHandle}
        onFocus={onFocus}
        onBlur={onBlur}
        testID={testID}>
        {state.routes.map(
          (route: NavigationRoute<ParamListBase, string>, index: number) => {
            const { title, drawerLabel, drawerIcon } =
              descriptors[route.key].options;
            return (
              <DrawerItem
                key={route.key}
                label={title || (drawerLabel as string) || route.name}
                icon={drawerIcon as any}
                isSelected={selectedItem === index}
                isFocused={focusedItem === index}
                drawerWidth={drawerWidth}
                isDrawerInFocus={isDrawerInFocus}
                onPress={onItemPress}
                routeName={route.name}
              />
            );
          },
        )}
      </TouchableOpacity>
    </TVFocusGuideView>
  );
};

const drawerItemDataUnchanged = (
  prevProps: DrawerItemListProps,
  nextProps: DrawerItemListProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(DrawerItemList, drawerItemDataUnchanged);

const styles = StyleSheet.create({
  listContainer: {
    flex: 1,
    justifyContent: 'center',
  },
});
