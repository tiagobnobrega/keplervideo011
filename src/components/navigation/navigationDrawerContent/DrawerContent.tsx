import { DrawerActions, useNavigation } from '@amzn/react-navigation__core';
import { DrawerContentComponentProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { isEqual } from 'lodash';
import React, { useState } from 'react';
import { ImageBackground, StyleSheet } from 'react-native';
import NavBackgroundGradient from '../../../assets/nav_drawer_background_gradient.png';
import { COLORS } from '../../../styles/Colors';
import { scaleUxToDp } from '../../../utils/pixelUtils';
import DrawerItemList from './DrawerItemList';
const drawerAverageWidth = scaleUxToDp(100);
const drawerExpandedWidth = scaleUxToDp(300);

const DrawerContent = (props: DrawerContentComponentProps) => {
  const navigation = useNavigation();

  const [isDrawerInFocus, setDrawerInFocus] = useState(false);
  const drawerWidth = isDrawerInFocus
    ? drawerExpandedWidth
    : drawerAverageWidth;

  const onFocusHandler = () => {
    setDrawerInFocus(true);
    navigation.dispatch(DrawerActions.openDrawer());
  };

  const onBlurHandler = () => {
    setDrawerInFocus(false);
    navigation.dispatch(DrawerActions.closeDrawer());
  };

  return (
    <ImageBackground
      source={NavBackgroundGradient}
      style={[styles.drawerBackground, { width: drawerWidth }]}>
      <DrawerItemList
        isDrawerInFocus={isDrawerInFocus}
        onDrawerListFocus={onFocusHandler}
        onDrawerListBlur={onBlurHandler}
        drawerWidth={drawerWidth}
        {...props}
      />
    </ImageBackground>
  );
};

const drawerDataUnchanged = (
  prevProps: DrawerContentComponentProps,
  nextProps: DrawerContentComponentProps,
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(DrawerContent, drawerDataUnchanged);

const styles = StyleSheet.create({
  drawerBackground: {
    height: '100%',
    backgroundColor: COLORS.DARK_GRAY,
    position: 'absolute',
  },
});
