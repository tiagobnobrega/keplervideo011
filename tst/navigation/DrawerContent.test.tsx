import { DrawerActions, useNavigation } from '@amzn/react-navigation__core';
import { fireEvent, render } from '@testing-library/react-native';
import { isEqual } from 'lodash';
import React from 'react';
import DrawerContent from '../../src/components/navigation/navigationDrawerContent/DrawerContent';
import DrawerItemList from '../../src/components/navigation/navigationDrawerContent/DrawerItemList';

jest.mock('@amzn/react-navigation__core', () => ({
  useNavigation: jest.fn(),
  DrawerActions: {
    openDrawer: jest.fn(),
    closeDrawer: jest.fn(),
  },
}));

const mockState = {
  index: 0,
  routes: [
    { key: 'home', name: 'Home' },
    { key: 'search', name: 'Search' },
    { key: 'settings', name: 'Settings' },
  ],
};

const mockDescriptors = {
  home: {
    options: {},
    navigation: {
      dispatch: jest.fn(),
    },
  },
  search: {
    options: {},
    navigation: {
      dispatch: jest.fn(),
    },
  },
  settings: {
    options: {},
    navigation: {
      dispatch: jest.fn(),
    },
  },
};

describe('DrawerContent Component', () => {
  let navigationMock: { dispatch: any };
  let mockProps: any;

  beforeEach(() => {
    navigationMock = {
      dispatch: jest.fn(),
    };
    (useNavigation as jest.Mock).mockReturnValue(navigationMock);
    jest.clearAllMocks();
  });

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  mockProps = {
    state: mockState,
    navigation: { dispatch: jest.fn() },
    descriptors: mockDescriptors,
  };

  it('renders correctly', () => {
    const component = render(<DrawerContent {...mockProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with', () => {
    const component = render(<DrawerContent {...mockProps} />);
    expect(component).toMatchSnapshot();
    expect(DrawerItemList).toBeDefined();
  });

  it('re-renders when props are not equal', () => {
    const { rerender } = render(<DrawerContent {...mockProps} />);
    rerender(<DrawerContent {...mockProps} />);
    expect(jest.mocked(isEqual)).toHaveBeenCalled();
  });
  it('should handle focus correctly', () => {
    const { UNSAFE_getByType } = render(<DrawerContent {...mockProps} />);

    const drawerItemList = UNSAFE_getByType(DrawerItemList);
    fireEvent(drawerItemList, 'onDrawerListFocus');

    expect(navigationMock.dispatch).toHaveBeenCalledWith(
      DrawerActions.openDrawer(),
    );
  });

  it('should handle blur correctly', () => {
    const { UNSAFE_getByType } = render(<DrawerContent {...mockProps} />);

    const drawerItemList = UNSAFE_getByType(DrawerItemList);
    fireEvent(drawerItemList, 'onDrawerListBlur');

    expect(navigationMock.dispatch).toHaveBeenCalledWith(
      DrawerActions.closeDrawer(),
    );
  });
});
