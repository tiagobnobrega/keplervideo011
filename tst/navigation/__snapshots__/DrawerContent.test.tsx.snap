// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerContent Component renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "backgroundColor": "#1C1C1C",
        "height": "100%",
        "position": "absolute",
      },
      Object {
        "width": 100,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/nav_drawer_background_gradient.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": 100,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`DrawerContent Component renders correctly with 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "backgroundColor": "#1C1C1C",
        "height": "100%",
        "position": "absolute",
      },
      Object {
        "width": 100,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/nav_drawer_background_gradient.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": 100,
        },
        undefined,
      ]
    }
  />
</View>
`;
